File: 0015-3sum/0015-3sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Set<List<Integer>> ans = new HashSet<>();
        for(int i=0; i< nums.length; i++){
            Map<Integer, Integer> dict = new HashMap<>();
            int target = -1* nums[i];
            for(int j = i+1; j < nums.length; j++){
                int diff = target - nums[j];
                if(!dict.containsKey(diff)){
                    dict.put(nums[j], j);
                }else{
                    List<Integer> appendList = new ArrayList<Integer>();
                    appendList.add(nums[j]);
                    appendList.add(nums[dict.get(diff)]);
                    appendList.add(nums[i]);

                    Collections.sort(appendList);
                    ans.add(appendList);
                }
            }
        }
        return new ArrayList<>(ans);
    }
}

File: 0015-3sum/0015-3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ans = []
        nums.sort()
        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue 
            target = -nums[i]
            j = i+1
            k = len(nums)-1
            while(j < k):
                sum = nums[j] + nums[k]
                if sum < target:
                    j+=1
                elif sum > target:
                    k-=1
                else:
                    ans.append([nums[i], nums[j], nums[k]])
                    while j < k and nums[j] == nums[j+1]:
                        j+=1
                    while j < k and nums[k] == nums[k-1]:
                        k-=1
                    j+=1
                    k-=1
        return ans
        